dotnet-desktop-guide/net/wpf/events/weak-event-patterns.md

```diff
@@ -0,0 +1,128 @@
+
+Listening for events can lead to memory leaks. The usual technique for listening to an event is to use language-specific syntax to attach a handler to an event on a source. For example, the C# statement `source.SomeEvent += new SomeEventHandler(MyEventHandler)` or the VB statement `AddHandler source.SomeEvent, AddressOf MyEventHandler`. However, this technique creates a strong reference from the event source to the event listener. Unless the event handler is explicitly unregistered, the object lifetime of the listener will be influenced by the object lifetime of the source. In certain circumstances, you might want the object lifetime of the listener to be controlled by other factors, such as whether it currently belongs to the visual tree of the application. Whenever the object lifetime of the source extends beyond the useful object lifetime of the listener, the listener is kept alive longer than necessary. In this case, the unallocated memory amounts to a memory leak.
+
+The weak event pattern is designed to solve the memory leak problem. The weak event pattern can be used when a listener needs to register for an event, but the listener doesn't explicitly know when to unregister. The weak event pattern can also be used when the object lifetime of the source exceeds the useful object lifetime of the listener. In this case, *useful* is determined by you. The weak event pattern allows the listener to register for and receive the event without affecting the object lifetime characteristics of the listener in any way. In effect, the implied reference from the source doesn't determine whether the listener is eligible for garbage collection. The reference is a weak reference, thus the naming of the weak event pattern and the related APIs. The listener can be garbage collected or otherwise destroyed, and the source can continue without retaining noncollectible handler references to a now destroyed object.
+
+## Who should implement the weak event pattern?
+
+The weak event pattern is primarily relevant to control authors. As a control author, you're largely responsible for the behavior and containment of your control and the impact it has on the applications in which it's inserted. This includes the control's object lifetime behavior, in particular the handling of the described memory leak problem.
+
+Certain scenarios inherently lend themselves to the application of the weak event pattern. One such scenario is data binding. In data binding, it's common for the source object to be independent of the listener object, which is a target of a binding. Many aspects of WPF data binding already have the weak event pattern applied in how the events are implemented.
+
+## How to Implement the weak event pattern
+
+There are three ways to implement the weak event pattern. The following table lists the three approaches and provides some guidance for when to use each.
+
+| Approach | When to Implement |
+|----------|-------------------|
```

Comment:
Let me know if you want the :::column format. It's not recommended, but I'm using it on another project because it scales images properly when columns are located inside tabs. In this text-only table, I'd recommend staying with standard markdown.